capella2polarion/connectors/polarion_worker.py

```diff
@@ -425,36 +461,104 @@ def compare_and_update_work_items(
+    ):
         """Create new documents."""
-        self.client.project_client.documents.create(documents)
+        client = self._get_client(document_project)
+        documents, _ = self._process_document_datas(client, document_datas)
+
+        client.documents.create(documents)
 
-    def update_documents(self, documents: list[polarion_api.Document]):
+    def update_documents(
+        self,
+        document_datas: list[data_models.DocumentData],
+        document_project: str | None = None,
+    ):
         """Update existing documents."""
-        self.client.project_client.documents.update(documents)
+        client = self._get_client(document_project)
+        documents, headings = self._process_document_datas(
+            client, document_datas
+        )
+
+        client.work_items.update(headings)
+        client.documents.update(documents)
+
+    def _process_document_datas(
+        self,
+        client: polarion_api.ProjectClient,
+        document_datas: list[data_models.DocumentData],
+    ):
+        documents = []
+        headings = []
+        for document_data in document_datas:
+            headings += document_data.headings
+            documents.append(document_data.document)
+            if document_data.text_work_item_provider.new_text_work_items:
+                self._create_and_update_text_work_items(
+                    document_data.text_work_item_provider.new_text_work_items,
+                    client,
+                )
+                document_data.text_work_item_provider.insert_text_work_items(
+                    document_data.document,
+                )
+        return documents, headings
 
     def get_document(
-        self, space: str, name: str
+        self, space: str, name: str, document_project: str | None = None
     ) -> polarion_api.Document | None:
         """Get a document from polarion and return None if not found."""
+        client = self._get_client(document_project)
         try:
-            return self.client.project_client.documents.get(
+            return client.documents.get(
                 space, name, fields={"documents": "@all"}
             )
         except polarion_api.PolarionApiBaseException as e:
             if e.args[0] == 404:
                 return None
             raise e
 
-    def update_work_items(self, work_items: list[polarion_api.WorkItem]):
-        """Update the given workitems without any additional checks."""
-        self.client.project_client.work_items.update(work_items)
-
     def load_polarion_documents(
-        self, document_paths: t.Iterable[tuple[str, str]]
-    ) -> dict[tuple[str, str], polarion_api.Document | None]:
-        """Load the given document references from Polarion."""
+        self,
+        document_infos: t.Iterable[data_models.DocumentInfo],
+    ) -> dict[
+        tuple[str | None, str, str],
+        tuple[polarion_api.Document | None, list[polarion_api.WorkItem]],
+    ]:
```

Comment:
"Good code reads like well written prose." - Robert C. Martin

We could make this more readable with a nice `type.Alias`. Even from looking in the return statement I can't quickly see what is the return value.